File: Leetcode/PY/Leetcode1005-K次取反后最大化的数组和.py
class Solution:
    def largestSumAfterKNegations(self, nums: List[int], k: int) -> int:
        mp = defaultdict(int)
        mn = int(1000)
        for num in nums:
            mp[num] += 1
            mn = min(mn,abs(num))
        for i in range(-100,0):
            if k == 0: break
            while k > 0 and i in mp and mp[i] != 0:
                mp[i] -= 1
                mp[-i] += 1
                k -= 1
        res = 0
        for i in range(-100,101):
            if i in mp and mp[i] > 0: res += mp[i] * i

        if k & 1 == 1: res -= 2 * mn
        return res

File: Leetcode/JS/Leetcode1005-K次取反后最大化的数组和.js
var largestSumAfterKNegations = function(nums, k) {
    let cnt = new Map();
    let mn = 1e9;
    for(const num of nums){
        cnt.set(num,(cnt.get(num) || 0) + 1);
        mn = Math.min(mn,Math.abs(num));
    }
    for(let i = -100;i < 0 && k;i++){
        while(k && cnt.has(i) && cnt.get(i) > 0){
            cnt.set(i,cnt.get(i)- 1);
            cnt.set(-i,(cnt.get(-i) || 0) + 1);
            k--;
        }
    }
    let res = 0;
    for(let i = -100;i <= 100;i++){
        if(cnt.has(i)){
            res += cnt.get(i) * i;
        }
    }
    if(k & 1) res -= 2 * mn;
    return res;
};